### PR TITLE
Add login decorator to improve tests

### PR DESCRIPTION
**Pull request by PhilippMatthes**
Closes: #284

### CODE REVIEW

**Comment by martinmo** [comment]
Nice idea, but does it really improve the code? The number of lines actually increased. But I understand the intension. Maybe the `assertTrue(self.client.login(...))` should be moved into the `setUp()` method where applicable?

**Comment by martinmo** [comment]
For this to work the test classes should be grouped into "logged in" and "not logged in" tests, of course.

**Comment by PhilippMatthes** [comment]
Maybe both? First, group all tests that need login into a separate test class. Second, add the decorator where applicable on `setUp` methods. The main reason for me to use a decorator is that, due to currently open issues, we will ultimately include more tests in the future that need a preceding login. It is far more convenient *and rememberable* (therefore reducing development overhead) to use this decorator than just pasting 
```python
self.assertTrue(self.client.login(username="bob", password="secret")) 
```
before the control flow. It focuses the control flow of these tests more on the essential parts of the unit tests and therefore also increases readability where applicable.

**Comment by martinmo** [comment]
I see your point. But two things speak against the decorator IMO:
- We remove the `self.assertTrue(...)` line and add `@assert_login(...)` => same amount of boilerplate.
- `self.assertTrue(...)` is more explicit. Have you checked the error message and stacktrace that is procuced when the decorator is used?

**Comment by PhilippMatthes** [comment]
Here is a sample error message produced by the decorator:
```
======================================================================
FAIL: test_redirect_to_slug (tests.tasks.test_views.TaskDetailViewTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/philippmatthes/Desktop/inloop/tests/decorators.py", line 12, in wrapper
    test.assertTrue(test.client.login(username=username, password=password))
AssertionError: False is not true

----------------------------------------------------------------------
```
Currently, I see why there is the same amount of boilerplate. However, we could extend the decorator to always perform the following steps instead of just one:
1. Validate that the given username exists. If not, throw a descriptive error.
2. Actually login and throw another descriptive error on failure.

**Comment by PhilippMatthes** [comment]
With the changes I just included, the error message for
```python
@assert_login("hello", "secret")
```
throws a more meaningful error message:
```
FAIL: test_task_visibility (tests.tasks.test_views.IndexViewTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/philippmatthes/Desktop/inloop/tests/decorators.py", line 15, in wrapper
    test.fail("There is no user with the username: {}".format(username))
AssertionError: There is no user with the username: hello
```
which is clearly an advantage over the current workflow IMHO.

**Comment by martinmo** [comment]
I am still not convinced if a method decorator improves this. Nevertheless, I have some improvements for the decorator code which I'll commit in a minute. Otherwise, although this is really interesting, let's try to not deviate too far and focus on some more important issues. I'll leave this PR open for the time being, I don't want to decide on this yet.

File: tests/decorators.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+    The decorated method must be in a subclass of django.test.TestCase.
+    """
+    def decorator(decorated_method):
+        @wraps(decorated_method)
+        def wrapper(self):
+            if not self.client.login(username=username, password=password):
+                raise AssertionError("Could not login with %s" % {'username': username,
+                                                                  'password': password})
+            return decorated_method(self)
```
Comment (martinmo) [inline tests/decorators.py]: At the moment, this decorator doesn't allow for arbitrary method arguments of the decorated method.

**Comment by martinmo** [comment]
Output in a failure scenario now looks like this:
```
FAIL: test_only_owner_can_access (tests.solutions.test_views.SolutionStatusViewTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/martin/Code/inloop/inloop/tests/decorators.py", line 24, in wrapper
    'password': password})
AssertionError: Could not login with {'username': 'alice', 'password': 'wr0ng'}
```

**Pull request closed without merging**